### PR TITLE
change Cache to use a generic comparable for keys [BREAKING]

### PR DESCRIPTION
**Pull request by amery**
groupcache can only use strings though.

## Summary by CodeRabbit

- **New Features**
	- Introduced generic type parameters for `Cache`, `Getter`, `Setter`, `Store`, and `SingleFlight` structures, enhancing type safety and flexibility for various key types.
	- Updated methods across packages to support generic keys, allowing for a broader range of key types while maintaining backward compatibility with string keys.

- **Bug Fixes**
	- Improved type safety in cache operations, reducing potential runtime errors related to type mismatches.

- **Documentation**
	- Enhanced clarity in method signatures and type definitions to reflect the new generic implementations.

### CODE REVIEW

File: x/memcache/singleflight.go [hunk 1]
```diff
@@ -237,21 +237,21 @@ type outreacher struct {
 }
 
 // Expire returns the error set by a failed outward.Get()
-func (p *outreacher) Err() error { return p.err }
+func (p *outreacher[K]) Err() error { return p.err }
 
 // Expire returns the data set by a successful outward.Get()
-func (p *outreacher) Bytes() []byte { return p.b }
+func (p *outreacher[K]) Bytes() []byte { return p.b }
```
[the remark this answers is not in the document]
Comment (karasz) [reply x/memcache/singleflight.go:243]: @amery can you fix this please?

[the remark this answers is not in the document]
Comment (amery) [reply x/memcache/singleflight.go]: amended

File: x/groupcache/groupcache.go [hunk 2]
```diff
@@ -15,10 +15,10 @@ import (
...
 
 var (
-	_ cache.Store  = (*HTTPPool)(nil)
-	_ http.Handler = (*HTTPPool)(nil)
-	_ cache.Store  = (*Pool)(nil)
-	_ cache.Cache  = (*Group)(nil)
+	_ cache.Store[string] = (*HTTPPool)(nil)
+	_ http.Handler        = (*HTTPPool)(nil)
+	_ cache.Store[string] = (*Pool)(nil)
+	_ cache.Cache[string] = (*Group)(nil)
```
[the remark this answers is not in the document]
Comment (amery) [reply x/groupcache/groupcache.go:21]: this groupcache package is a shim over github.com/mailgun/groupcache/v2

File: x/memcache/singleflight.go [hunk 3]
```diff
@@ -11,42 +11,42 @@ import (
 )
 
 var (
-	_ cache.Getter = (*SingleFlight)(nil)
-	_ cache.Setter = (*SingleFlight)(nil)
+	_ cache.Getter[string] = (*SingleFlight[string])(nil)
+	_ cache.Setter[string] = (*SingleFlight[string])(nil)
```
[the remark this answers is not in the document]
Comment (amery) [reply x/memcache/singleflight.go:15]: added now

**Review by karasz — APPROVED** [review]
LGTM

**Comment by amery** [comment]
`darvaza.org/cache@v0.3.0` tagged

**Comment by amery** [comment]
* `darvaza.org/cache/x/groupcache@v0.2.0`
* `darvaza.org/cache/x/memcache@v0.2.0`
* `darvaza.org/cache/x/protosink@v0.1.0`

**Pull request merged**